### PR TITLE
use `towncrier` to handle changelog entries

### PR DESCRIPTION
**Pull request by zacharyburnett**
using `towncrier` to handle changelog entries will
- drastically reduce (or even eliminate) branch conflicts in `CHANGES.rst` we currently experience with PRs
- keep the sections in `CHANGES.rst` consistent and eliminate duplicate sections
- allow automatic links to be generated in the changelog entries

[image](https://roman-pipeline--1375.org.readthedocs.build/en/1375/roman/changes.html)

`towncrier` expects "news fragment" files (text files in the `changes/` directory with filenames in the format `<PR#>.<changetype>.rst`, i.e. for this PR it would be `changes/1375.docs.rst`). See docs at https://towncrier.readthedocs.io/en/latest/tutorial.html#creating-news-fragments

when ready to make a release, run `towncrier build` to ingest the news fragments and generate a changelog section in `CHANGES.rst` with all the new change log entries for that release (this clears the `changes/` directory of all news fragment files). This step should either be done before making a release, or could probably be added to a GitHub workflow triggered on release (to insert a commit and remake the tag).

After merging this PR the `Release Process` wiki page will need to be updated to include a step to run `towncrier build` instead of manually editing `CHANGES.rst`

**Checklist**
- [x] added entry in `CHANGES.rst` under the corresponding subsection
- [N/A] ~updated relevant tests~
- [x] updated relevant documentation
- [x] updated relevant milestone(s)
- [x] added relevant label(s)
- [N/A] ~ran regression tests, post a link to the Jenkins job below. [How to run regression tests on a PR](https://github.com/spacetelescope/romancal/wiki/Running-Regression-Tests-Against-PR-Branches)~

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -227,3 +227,76 @@ archs = [
...
+name = "`skymatch`"
+
+[tool.towncrier.fragment.source_catalog]
+name = "`source_catalog`"
+
+[tool.towncrier.fragment.stpipe]
+name = "`stpipe`"
+
+[tool.towncrier.fragment.tweakreg]
+name = "`tweakreg`"
```
Comment (zacharyburnett) [inline pyproject.toml:302]: These are the changelog subsections that `towncrier` generates (specified by the change type in the filename of the news fragment). I copied them from the labels made in `.github/labeler.yml`; do we want to update these to match the step names?

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -9,7 +9,35 @@ Closes #
 This PR addresses ...
 
 **Checklist**
-- [ ] added entry in `CHANGES.rst` under the corresponding subsection
+- [ ] for a public change, added a [towncrier news fragment](https://towncrier.readthedocs.io/en/stable/tutorial.html#creating-news-fragments) <details><summary>`changes/<PR#>.<changetype>.rst`</summary>
```
Comment (zacharyburnett) [inline .github/pull_request_template.md:12]: **Checklist**
 - [ ] for a public change, added a [towncrier news fragment](https://towncrier.readthedocs.io/en/stable/tutorial.html#creating-news-fragments) <details><summary>`changes/<PR#>.<changetype>.rst`</summary>

     - ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
     - ``changes/<PR#>.docs.rst``
     - ``changes/<PR#>.stpipe.rst``
     - ``changes/<PR#>.associations.rst``
     - ``changes/<PR#>.scripts.rst``

     ## steps
     - ``changes/<PR#>.dq_init.rst``
     - ``changes/<PR#>.saturation.rst``
     - ``changes/<PR#>.refpix.rst``
     - ``changes/<PR#>.linearity.rst``
     - ``changes/<PR#>.dark_current.rst``
     - ``changes/<PR#>.ramp_fitting.rst``
     - ``changes/<PR#>.assign_wcs.rst``
     - ``changes/<PR#>.flatfield.rst``
     - ``changes/<PR#>.photom.rst``
     - ``changes/<PR#>.source_detection.rst``
     - ``changes/<PR#>.tweakreg.rst``
     - ``changes/<PR#>.flux.rst``
     - ``changes/<PR#>.jump_detection.rst``
     - ``changes/<PR#>.mosaic_pipeline.rst``
     - ``changes/<PR#>.outlier.rst``
     - ``changes/<PR#>.patch_match.rst``
     - ``changes/<PR#>.resample.rst``
     - ``changes/<PR#>.skymatch.rst``
     - ``changes/<PR#>.source_catalog.rst``
   </details>
 - [ ] updated relevant tests
 - [ ] updated relevant documentation
 - [ ] updated relevant milestone(s)

**Comment by zacharyburnett** [comment]
current changelog (`towncrier build --draft`):
```rst
0.16.2.dev2+g482ef5c7 (2024-08-27)
==================================

Documentation
-------------

- handle changelog entries with `towncrier` (`#1375
  <https://github.com/spacetelescope/romancal/1375>`_)


``mosaic_pipeline``
-------------------

- Only load patch table when needed. (`#1367
  <https://github.com/spacetelescope/romancal/1367>`_)
```

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -9,7 +9,35 @@ Closes #
...
+    - ``changes/<PR#>.source_detection.rst``
+    - ``changes/<PR#>.tweakreg.rst``
+    - ``changes/<PR#>.skymatch.rst``
+    - ``changes/<PR#>.outlier_detection.rst``
+    - ``changes/<PR#>.resample.rst``
+    - ``changes/<PR#>.source_catalog.rst``
+  </details>
 - [ ] updated relevant tests
 - [ ] updated relevant documentation
 - [ ] updated relevant milestone(s)
```
Comment (zacharyburnett) [inline .github/pull_request_template.md:43]: **Checklist**
 - [ ] for a public change, added a towncrier news fragment in `changes/` <details><summary>`echo "changed something" > changes/<PR#>.<changetype>.rst`</summary>

     - ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
     - ``changes/<PR#>.docs.rst``
     - ``changes/<PR#>.stpipe.rst``
     - ``changes/<PR#>.associations.rst``
     - ``changes/<PR#>.scripts.rst``
     - ``changes/<PR#>.mosaic_pipeline.rst``
     - ``changes/<PR#>.patch_match.rst``

     ## steps
     - ``changes/<PR#>.dq_init.rst``
     - ``changes/<PR#>.saturation.rst``
     - ``changes/<PR#>.refpix.rst``
     - ``changes/<PR#>.linearity.rst``
     - ``changes/<PR#>.dark_current.rst``
     - ``changes/<PR#>.jump_detection.rst``
     - ``changes/<PR#>.ramp_fitting.rst``
     - ``changes/<PR#>.assign_wcs.rst``
     - ``changes/<PR#>.flatfield.rst``
     - ``changes/<PR#>.photom.rst``
     - ``changes/<PR#>.flux.rst``
     - ``changes/<PR#>.source_detection.rst``
     - ``changes/<PR#>.tweakreg.rst``
     - ``changes/<PR#>.skymatch.rst``
     - ``changes/<PR#>.outlier_detection.rst``
     - ``changes/<PR#>.resample.rst``
     - ``changes/<PR#>.source_catalog.rst``
   </details>
 - [ ] updated relevant tests
 - [ ] updated relevant documentation
 - [ ] updated relevant milestone(s)

**Comment by schlafly** [comment]
Do you want to talk about this at the tag up tomorrow (3 pm)?

**Comment by zacharyburnett** [comment]
> Do you want to talk about this at the tag up tomorrow (3 pm)?

sure, where is it?

**Review by schlafly — APPROVED** [review]
Thanks Zach, LGTM.

**Pull request merged**